File: 1028-interval-list-intersections/interval-list-intersections.cpp
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList, vector<vector<int>>& secondList) {
        vector<vector<int>> ans;
        int p1 = 0, p2 = 0;
        while (p1 < firstList.size() || p2 < secondList.size()) {
            if (p1 != firstList.size() && p2 != secondList.size()) {
                if (max(firstList[p1][0], secondList[p2][0]) <= min(firstList[p1][1], secondList[p2][1]))
                    ans.push_back({max(firstList[p1][0], secondList[p2][0]), 
                                min(firstList[p1][1], secondList[p2][1])});
                if (firstList[p1][1] > secondList[p2][1])
                    p2++;
                else
                    p1++;
            }
            else if (p1 != firstList.size() && p2 == secondList.size()) 
                p1++;
            else
                p2++;
        }
        return ans;
    }
};

File: 1028-interval-list-intersections/interval-list-intersections.go
func intervalIntersection(firstList [][]int, secondList [][]int) [][]int {
    p1, p2 := 0, 0
    res := [][]int{}

    for p1 < len(firstList) && p2 < len(secondList) {
        if isJoin(firstList[p1], secondList[p2]) {
            res = append(res, getIntersection(firstList[p1], secondList[p2]))
        }
        
        if MinSecondPoint(firstList[p1], secondList[p2]) {
            p1++
        } else {
            p2++
        }
    }

    return res
}

func MinSecondPoint(firstList, secondList []int) bool {
    return firstList[1] < secondList[1]
}

func isJoin(firstList, secondList []int) bool {
    return max(firstList[0], secondList[0]) <= min(firstList[1], secondList[1])
}

func getIntersection(firstList, secondList []int) []int {
    return []int{max(firstList[0], secondList[0]), min(firstList[1], secondList[1])}
}